iavl/iavl_proof2.go

```diff
@@ -167,52 +171,266 @@ func (proof *KeyNotExistsProof) Verify(key []byte, root []byte) error {
 	return nil
 }
 
-type KeyRangeExistsProof struct {
-	RootHash   data.Bytes
-	PathToKeys []*PathToKey
+// KeyRangeProof is proof that a range of keys does or does not exist.
+type KeyRangeProof struct {
+	RootHash   data.Bytes   `json:"root_hash"`
+	PathToKeys []*PathToKey `json:"paths"`
+
+	LeftPath *PathToKey        `json:"left_path"`
+	LeftNode IAVLProofLeafNode `json:"left_node"`
+
+	RightPath *PathToKey        `json:"right_path"`
+	RightNode IAVLProofLeafNode `json:"right_node"`
+}
+
+func (proof *KeyRangeProof) String() string {
+	paths := []string{}
+	for _, p := range proof.PathToKeys {
+		paths = append(paths, "\t"+p.String())
+	}
+	inner := fmt.Sprintf("PathToKeys: \n\t%s", strings.Join(paths, ",\n"))
+	inner += fmt.Sprintf("LeftNode: %#v (%x)\n", proof.LeftNode, proof.LeftNode.Hash())
+	inner += fmt.Sprintf("RightNode: %#v (%x)", proof.RightNode, proof.RightNode.Hash())
+	return "&KeyRangeProof{\n\t" + inner + "\n}"
 }
 
-func (proof *KeyRangeExistsProof) Verify(key []byte, value []byte, root []byte) bool {
-	return false
+// Returns a list of all paths, in order.
+func (proof *KeyRangeProof) paths() []*PathToKey {
+	paths := proof.PathToKeys[:]
+	if proof.LeftPath != nil {
+		paths = append([]*PathToKey{proof.LeftPath}, paths...)
+	}
+	if proof.RightPath != nil {
+		paths = append(paths, proof.RightPath)
+	}
+	return paths
 }
 
-func (node *IAVLNode) constructKeyExistsProof(t *IAVLTree, key []byte, proof *KeyExistsProof) ([]byte, error) {
+// Checks that all paths are adjacent to one another, with no gaps.
+func (proof *KeyRangeProof) verifyPathAdjacency() error {
+	paths := proof.paths()
+	for i := 0; i < len(paths)-1; i++ {
+		// Always check from left to right, since paths are always in ascending order.
+		if !paths[i].isLeftAdjacentTo(paths[i+1]) {
+			return errors.Errorf("paths #%d and #%d are not adjacent", i, i+1)
+		}
+	}
+	return nil
+}
+
+// Verify that a range proof is valid.
+func (proof *KeyRangeProof) Verify(
+	startKey, endKey []byte, limit int, keys, values [][]byte, root []byte,
+) error {
+	if len(proof.PathToKeys) != len(keys) || len(values) != len(keys) {
+		return errors.New("wrong number of keys or values for proof")
+	}
+	if len(proof.PathToKeys) == 0 && proof.LeftPath.isEmpty() && proof.RightPath.isEmpty() {
+		return errors.New("proof is incomplete")
+	}
+
+	// If startKey > endKey, reverse the keys and values, since our proofs are
+	// always in ascending order.
+	if bytes.Compare(startKey, endKey) == 1 {
+		startKey, endKey = endKey, startKey
+
+		ks := make([][]byte, len(keys))
+		vs := make([][]byte, len(keys))
+		for i, _ := range keys {
+			ks[len(ks)-1-i] = keys[i]
+			vs[len(vs)-1-i] = values[i]
+		}
+		keys, values = ks, vs
+	}
+
+	if len(keys) > 0 {
```

Comment:
Ok, I made some changes: a15153ebe3b510973944cfc8702796ff8e7ed819 - there's much more that can be done by sharing code between the different proofs, but I'll save that for the next PR where I have the first/last-in-range proofs.